### PR TITLE
Drop Ruby 2.3/2.4/2.5 support; puppet_forge: Use 4.1 and newer

### PR DESCRIPTION
**Pull request by bastelfreak**

### CODE REVIEW

**Comment by justinstoller** [comment]
We created a 4.x branch for this kind of work and then de-prioritized the work we were going to land in it. Do you see anything in the 4.x branch that you would be opposed to us releasing? I think if you retargeted this at 4.x, then we renamed `main` to `3.x` and `4.x` to `main` and did a release we'd be good?

File: .github/workflows/rspec_tests.yml
Comment (justinstoller) [inline .github/workflows/rspec_tests.yml:18]: jruby 9.4 is out now and ships in Puppet Server 8 along with Ruby 3.2. It'd be great, if we're going to drop support for older versions we added support for newer ones. I'd love to see Ruby 3.1 & 3.2 in addition to JRuby 9.4 added to this matrix.

**Comment by justinstoller** [comment]
Mmmmm... Thinking about this more, I think we _can_ drop this support if we want to in 3.x, is it required?

I instinctively thought about moving to the latest releases when I saw the dependency changes, but I think that would require a bump to 4.0, which I suppose isn't really what you were going for, eh?

FWIW, I'd love to see us do a 4.0 release with full Puppet 8 support this summer.

File: .github/workflows/rspec_tests.yml
Comment (bastelfreak) [reply .github/workflows/rspec_tests.yml:18]: yes. But that's an enhancement and not related to this breaking change, so I wanted to provide that in another PR.

**Comment by bastelfreak** [comment]
> FWIW, I'd love to see us do a 4.0 release with full Puppet 8 support this summer.

Yes please! I'm waiting for this since over a year :)

I retargeted to 4.x. But main has newer commits that are missing in 4.x. I will create another PR to backport it.

**Comment by bastelfreak** [comment]
I raised https://github.com/puppetlabs/r10k/pull/1345 to backport changes from main to 4.x. When that's merged I can rebase #1336

**Comment by bastelfreak** [comment]
@Magisus @justinstoller what can I do to move this forward?

**Comment by justinstoller** [comment]
Hey! Sorry, I dropped the ball on this, Tim. I've reviewed all the applicable Dependabot PRs from your last PR and got an improvement Charlie wanted in merged for the 3.x series. I'm currently fighting our CI to get those through and unblock your work. 

I'll update with my progress tomorrow!

**Comment by bastelfreak** [comment]
Hi Justin, thanks for picking this up again. Let me know if I can help.

**Comment by justinstoller** [comment]
Still fighting internal issues related to some testing dependency updates.

**Comment by justinstoller** [comment]
Thanks for your patience on this, Tim. I've pushed up my local changes that _I_ think will unblock us, but I need to confirm when I get back from vacation on Thursday. https://github.com/puppetlabs/r10k/pull/1349

**Comment by justinstoller** [comment]
Sorry this hasn't moved forward yet. We _have_ been working on it off and on this whole time but we're tied up trying to get some testing infrastructure updated (we're hoping to get this onto an internal test fixture before merging my linked PR https://github.com/puppetlabs/puppetlabs-xinetd/pull/125)

**Comment by justinstoller** [comment]
jeez, that took _forever_. https://github.com/puppetlabs/r10k/pull/1349 is merged and once that's gone through our internal CI we'll get all this moving again. Probably Thursday.

**Review by justinstoller — APPROVED** [review]

**Comment by justinstoller** [comment]
there's a few more acceptance tests I need to do but I'm _incredibility_ sick of blocking on them. I will still work on fixing them up but want to get these in and do a final release of 3.x first thing next week with a 4.0 shortly after. :crossed_fingers:

**Pull request merged**
